### ISSUE TITLE
make SDS works E2E in Istio 1.1 branch

### ISSUE THREAD

**Issue by quanjielin** [istio/istio#9035]
Use this bug to track the work for [secret discovery service(SDS)](https://www.envoyproxy.io/docs/envoy/latest/configuration/secret#config-secret-discovery-service) works E2E in master branch.

The SDS change on  ```collab-gcp-identity``` is almost merged into master now, next step is making SDS E2E flow work in master.

Workitems for SDS to work in master(based on ```collab-gcp-identity```): 
- support multiple CA :  fetching root CA through SDS instead of baking root CA into proxy docker image like in ```collab-gcp-identity``` today
- accommodate envoy latest changes: In SDS config(distributed from pilot), channel credential required to be ```local_credential``` otherwise token cannot be passed through. 

Then change on e2e flow to support for k8s sa: 
- pilot side change: construct envoy config so that envoy could fetch/send out k8s sa jwt(trustworthy jwt preferred, normal k8s sa jwt if trustworthy not available) token in SDS request header.
- sidecar injector: inject token volume mount in istio-proxy container, both for manual and auto webhook
- nodeagent side: parse k8s jwt token, authN provider specific plugin(exchange token with IAM Service Account Credentials API), CSR to CAs, CA adapter for pluggable CA.
- citadel side: validate jwt(trustworthy or normal jwt) sent by nodeagent and issue cert. 
- integration test, mTLS works through identity provisioned by SDS flow. 
- documentation.

**Comment by quanjielin** [comment]
change on ```local_credential``` is blocked by https://github.com/envoyproxy/envoy/issues/4578

**Comment by quanjielin** [comment]
1. tried k8s sa jwt token volume mount, the file is only accessible to root user, not istio-proxy(which envoy runs as); need follow up. 
2. tried plugin call cred(to workaround issue 1, create fake token file in docker image file to avoid permission issue), didn't see k8s sa jwt pass through from envoy to nodeagent, https://github.com/envoyproxy/envoy/issues/4630

**Comment by quanjielin** [comment]
need to intro dependent https://github.com/google/go-genproto/blob/master/googleapis/iam/credentials/v1/common.pb.go

**Comment by quanjielin** [comment]
To inject token volume mount, need to update SHA istio/vendor/k8s.io/api/, current [VolumeProjection](https://github.com/istio/istio/blob/master/vendor/k8s.io/api/core/v1/types.go#L1690) miss [ServiceAccountToken](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L1575) which is added to k8s recently, otherwise the result of injection is like below

```
      volumes:
      - name: istio-token
        projected:
          sources:
          - {}
```
expected:
```
- name: istio-token
  projected:
    sources:
    - serviceAccountToken:
        path: istio-token
        expirationSeconds: 3600
        audience: istio
```

**Comment by quanjielin** [comment]
hit below error when nodeagent initialize CA client to calls into citadel 
```
quanlin@quanlin:~/integration/1220/citadel$ kubectl logs nodeagent-znwmc nodeagent -n istio-system
2018-12-20T18:05:11.898453Z	info	unalbe to fetch CA TLS root cert, retry in 2s
2018-12-20T18:05:13.898803Z	info	unalbe to fetch CA TLS root cert, retry in 2s

```

**Comment by quanjielin** [comment]
https://github.com/istio/istio/issues/10283

**Comment by quanjielin** [comment]
for some reason, app deployed by sidecar auto injection webhook doesn't have k8s normal jwt in ```/var/run/secrets/kubernetes.io/serviceaccount```(manually injection works fine), need to figure out why

**Comment by costinm** [comment]
It never had it - it's one of the reasons we stopped making k8s calls from
sidecar when auto-injection was added.
Did you test the new token type (with audience ) in auto-injected pods ? I
suspect it behaves the same.

As I mentioned before - I believe using the CNI plugin to handle
communication and auth to the node agent is the
most reliable ( and secure ) option.

**Comment by quanjielin** [comment]
@costinm, I verified trustworthy jwt was mounted to proxy container(```/var/run/secrets/tokens/ istio-token```) last Friday with fix https://github.com/istio/istio/pull/10636

**Comment by quanjielin** [comment]
verified workload provision pipeline envoy -> nodeagent -> citadel works, mtls works.

**Comment by jasminejaksic** [comment]
Per Oliver this is working, closing this issue

**Comment by quanjielin** [comment]
Remove this issue from 1.1 radar, as envoy->nodeagent->citadel path works.
Keep it open until Google CA e2e works(which is the original intention for this issue)

**Comment by hzxuzhonghu** [comment]
@quanjielin Where can I find more doc about how to use SDS?

**Comment by quanjielin** [comment]
@hzxuzhonghu we will add documentation in istio.io soon, for now you could refer https://github.com/istio/istio/pull/10605 and https://github.com/istio/istio/pull/10887 (for integration test for sds e2e)

**Comment by hzxuzhonghu** [comment]
Great, thanks.

**Comment by quanjielin** [comment]
This is done
